components/license-control-service/pkg/storage/storage.go

```diff
@@ -0,0 +1,261 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	_ "github.com/lib/pq" // required for pg backend
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/components/license-control-service/pkg/keys"
+	"github.com/chef/automate/lib/db/migrator"
+	"github.com/chef/automate/lib/logger"
+)
+
+// NoLicenseError is the error returned by a backend when it does not
+// have a configured License.
+type NoLicenseError struct{ backend string }
+
+func (n *NoLicenseError) Error() string {
+	return fmt.Sprintf("No license found in %s storage backend", n.backend)
+}
+
+// NotInitializedError is returned by a CurrentBackend when the user fails to
+// call Init() before interacting with it.
+type NotInitializedError struct{ backend string }
+
+func (n *NotInitializedError) Error() string {
+	return fmt.Sprintf("Init() never called on %s storage backend", n.backend)
+}
+
+// A CurrentBackend can be used as a storage backend for the
+// license-control-service.
+type CurrentBackend interface {
+	// Init migrates data from any known Upgradeable Backends
+	Init(context.Context, *keys.LicenseParser) error
+	// GetLicense returns the currently configured autoamte
+	// license. If no license is configured, NoLicense is
+	// returend.
+	GetLicense(context.Context) (string, keys.LicenseMetadata, error)
+	// SetLicense stores the given license in the backend.
+	SetLicense(context.Context, string) error
+}
+
+// An UpgradeableBackend is an old backend that can no longer be used
+// as a backend for the license-control-service but which might have
+// data that needs to be migrated into the CurrentBackend.
+type UpgradeableBackend interface {
+	// GetLicense returns the currently configured autoamte
+	// license. If no license is configured, ErrNoLicense is
+	// returend.
+	GetLicense(context.Context) (string, error)
+	// Cleanup removes the state kept by this backend. After
+	// Cleanup is called, GetLicense should return a
+	// NoLicenseError.
+	Cleanup(context.Context) error
+}
+
+var _ CurrentBackend = &PGBackend{}
+var _ UpgradeableBackend = &FileBackend{}
+
+func NewCurrentBackend(pgURL string, migrationPath string, legacyFilePath string) CurrentBackend {
+	return &PGBackend{
+		pgURL:             pgURL,
+		migrationPath:     migrationPath,
+		legacyFileBackend: &FileBackend{path: legacyFilePath},
+	}
+}
+
+// PgBackend is the CurrentBackend used in production. It stores the
+// raw key data in PostgreSQL.
+type PGBackend struct {
+	pgURL             string
+	migrationPath     string
+	legacyFileBackend UpgradeableBackend
+
+	db          *sql.DB
+	initialized bool
+}
+
+func (p *PGBackend) Init(ctx context.Context, l *keys.LicenseParser) error {
+	err := p.doInit(ctx, l)
```

Comment:
Hmm could use `sync.Once`, too, I guess? But probably doesn't matter since we ensure this is only called once? I guess every handler could call init, and it would only happen on the first run, and we wouldn't have to worry about "not initialized" conditions? 🤔

Reply:
Yeah, to be honest I'm not sure how much any of it is needed.  The goal was to make sure the callers always called Init() before any of the other methods.

I suppose a different way to do that would be to define

```
type UnintializedCurrentBackend interface {
    Init(context.Context, *keys.LicenseParser) (CurrentBackend, error)
}
```

and then the caller could only get a CurrentBackend via Init()...

Reply:
🤔 That's a nice idea. 💡 

But really, I wonder how much we try to foresee our own future mistakes -- I'm not sure we actually build precautions in the right places...

Reply:
Right, in this case, you still need tests that will fail if this error is hit and if you have that test, you can just have a test that fails on the inevitable error that you'll get if GetLicense is called before Init().

I think I'll remove it, it'll also clean up a lot of code.